substrate/frame/contracts/uapi/src/lib.rs

```diff
@@ -0,0 +1,172 @@
+	/// Can only be returned from `call` and `instantiate`.
+	CalleeReverted = 2,
+	/// The passed key does not exist in storage.
+	KeyNotFound = 3,
+	/// Deprecated and no longer returned: There is only the minimum balance.
+	_BelowSubsistenceThreshold = 4,
+	/// Transfer failed for other not further specified reason. Most probably
+	/// reserved or locked balance of the sender that was preventing the transfer.
+	TransferFailed = 5,
+	/// Deprecated and no longer returned: Endowment is no longer required.
+	_EndowmentTooLow = 6,
+	/// No code could be found at the supplied code hash.
+	CodeNotFound = 7,
+	/// The account that was called is no contract.
+	NotCallable = 8,
+	/// The call to `debug_message` had no effect because debug message
+	/// recording was disabled.
+	LoggingDisabled = 9,
+	/// The call dispatched by `call_runtime` was executed but returned an error.
+	CallRuntimeFailed = 10,
+	/// ECDSA public key recovery failed. Most probably wrong recovery id or signature.
+	EcdsaRecoveryFailed = 11,
+	/// sr25519 signature verification failed.
+	Sr25519VerifyFailed = 12,
+	/// The `xcm_execute` call failed.
+	XcmExecutionFailed = 13,
+	/// The `xcm_send` call failed.
+	XcmSendFailed = 14,
+}
+
+/// The raw return code returned by the host side.
+#[repr(transparent)]
+pub struct ReturnCode(u32);
+
+/// Used as a sentinel value when reading and writing contract memory.
+///
+/// We use this value to signal `None` to a contract when only a primitive is
+/// allowed and we don't want to go through encoding a full Rust type.
+/// Using `u32::Max` is a safe sentinel because contracts are never
+/// allowed to use such a large amount of resources. So this value doesn't
+/// make sense for a memory location or length.
+const SENTINEL: u32 = u32::MAX;
+
+impl From<ReturnCode> for Option<u32> {
+	fn from(code: ReturnCode) -> Self {
+		(code.0 < SENTINEL).then_some(code.0)
+	}
+}
+
+impl ReturnCode {
+	/// Returns the raw underlying `u32` representation.
+	pub fn into_u32(self) -> u32 {
+		self.0
+	}
+	/// Returns the underlying `u32` converted into `bool`.
+	pub fn into_bool(self) -> bool {
+		self.0.ne(&0)
+	}
+}
+
+type Result = core::result::Result<(), ReturnErrorCode>;
+
+#[inline(always)]
+#[cfg(any(target_arch = "wasm32", target_arch = "riscv32"))]
+fn extract_from_slice(output: &mut &mut [u8], new_len: usize) {
+	debug_assert!(new_len <= output.len());
+	let tmp = core::mem::take(output);
+	*output = &mut tmp[..new_len];
+}
+
+#[inline(always)]
+#[cfg(any(target_arch = "wasm32", target_arch = "riscv32"))]
+fn ptr_and_len_from_slice(data: &mut Option<&mut [u8]>) -> (*mut u8, u32) {
+	match data {
+		Some(ref mut data) => (data.as_mut_ptr(), data.len() as _),
+		None => (SENTINEL as _, 0),
+	}
+}
+
+#[inline(always)]
+#[cfg(any(target_arch = "wasm32", target_arch = "riscv32"))]
+fn ptr_from_slice(data: &Option<&[u8]>) -> *const u8 {
+	match data {
+		Some(ref data) => data.as_ptr(),
+		None => SENTINEL as _,
+	}
+}
```

Comment:
moved it with this commit https://github.com/paritytech/polkadot-sdk/pull/2347/commits/b55b3706ca4966165bd3d9ea0c0b6d4a8f09f11c
Not sure if that's what you had in mind. You need one way or another to have these functions behind the cfg flag or the build will complain they are never used

Reply:
Just conditionally compile the whole `Api` module to avoid this problem. Then we don't need the `common` sub module and don't need to duplicate the `ApiImpl` type. Can just be defined once in the module root.